{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "00d93533",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import os\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from io import StringIO\n",
    "from Bio.Blast import NCBIXML"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "eecbc463",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_folder = './015_Lazazzara_10-03-2022'\n",
    "seq_folder = './sequences/'\n",
    "output_folder = './consensus/' \n",
    "edit_folder = './edited/'\n",
    "suffix = '.phd.1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "7d209e81",
   "metadata": {},
   "outputs": [],
   "source": [
    "# getting the sequences from the sanger output (.phd.1 in this case)\n",
    "fs = [f for f in os.listdir(input_folder) if f.endswith(suffix)] # collecting file names with sequencing information\n",
    "qualdic = {} # collectiong information on filename : [base quality scores]\n",
    "for f in fs:\n",
    "    name = '_'.join(f.split('_')[0:2]) # file name\n",
    "    sequence = '' # sequence variable\n",
    "    qs = [] # quality scores list\n",
    "    for row in open(input_folder+'/'+f).readlines()[103:-53]:\n",
    "        row = row.strip('\\n') # remove new lines \\n\n",
    "        row = row.split(' ') # make each row a list\n",
    "        sequence += row[0] # the first list component is the name of the base; add it to the sequence variable\n",
    "        qs.append(row[1]) # the second list component is the quality of the base; add it to the quality scores varaible\n",
    "    # conditions to set the sequence description\n",
    "    if '1' in name:\n",
    "        des = 'Arboridia Puglia'\n",
    "    elif '2' in name:\n",
    "        des = 'Empoasca decipiens'\n",
    "    elif '3' in name:\n",
    "        des = 'Arboridia Creta'\n",
    "    elif '4' in name:\n",
    "        des = 'Arboridia Dalmazia'\n",
    "    sequence = SeqRecord(Seq(sequence),id=name,description=des) # defining the SeqRecord object for the FASTA file write\n",
    "    # write FASTA file\n",
    "    with open(seq_folder+name+'.fa','w') as handle:\n",
    "        SeqIO.write(sequence,handle,'fasta')\n",
    "    qualdic[name] = qs # adding the filename : [base quality scores] to the disctionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "0b8c4ff4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequence(f,s=0,e=0):\n",
    "    f = SeqIO.read(f,'fasta')\n",
    "    if e != 0:\n",
    "        f.seq = f.seq[s:e]\n",
    "    return f"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "id": "2956897c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# funciton input are:\n",
    "# sequence1, sequence2, seq1_start, seq2_start, id seq1, id seq2, description, quality_dic, quality_check\n",
    "def get_consensus(s1, s2, s1_s, s2_s, id1, id2, qs_d, n): \n",
    "    name = \"\".join([c for c in id1 if c in id2]) # get the name of the consensus\n",
    "    q1 = qs_d[id1.split(' ')[0]] # get the quality of the sequence 1\n",
    "    q2 = qs_d[id2.split(' ')[0]][::-1] # get the quality of the sequence 2\n",
    "    s = ''\n",
    "    for i in range(len(s1)):\n",
    "        a = s1[i]\n",
    "        b = s2[i]\n",
    "        q_a = int(q1[s1_s+i])\n",
    "        q_b = int(q2[s2_s+i])\n",
    "        if a == b:\n",
    "            s += a\n",
    "        elif q_a > q_b and q_a > n:\n",
    "            print(id1,a,q_a,id2,b,q_b, 'a')\n",
    "            s += a\n",
    "        elif q_b > q_a and q_b > n:\n",
    "            print(id1,a,q_a,id2,b,q_b, 'b')\n",
    "            s += b\n",
    "        elif q_a > 20 and q_b > 20:\n",
    "            print(id1,a,q_a,id2,b,q_b)\n",
    "            if a == 'A':\n",
    "                if b == 'G':\n",
    "                    s += 'R'\n",
    "                elif b == 'T':\n",
    "                    s += 'W'\n",
    "                elif b == 'C':\n",
    "                    s += 'M'\n",
    "                else:\n",
    "                    s += 'N'\n",
    "            if a == 'G':\n",
    "                if b == 'A':\n",
    "                    s += 'R'\n",
    "                elif b == 'T':\n",
    "                    s += 'K'\n",
    "                elif b == 'C':\n",
    "                    s += 'S'\n",
    "                else:\n",
    "                    s += 'N'\n",
    "            if a == 'C':\n",
    "                if b == 'G':\n",
    "                    s += 'S'\n",
    "                elif b == 'T':\n",
    "                    s += 'Y'\n",
    "                elif b == 'A':\n",
    "                    s += 'M'\n",
    "                else:\n",
    "                    s += 'N'\n",
    "            if a == 'T':\n",
    "                if b == 'G':\n",
    "                    s += 'K'\n",
    "                elif b == 'A':\n",
    "                    s += 'W'\n",
    "                elif b == 'C':\n",
    "                    s += 'Y'\n",
    "                else:\n",
    "                    s += 'N'\n",
    "        else:\n",
    "            print(id1,a,q_a,id2,b,q_b,'N')\n",
    "            s += 'N'\n",
    "    if '1' in name:\n",
    "        des = 'Arboridia Puglia'\n",
    "    elif '2' in name:\n",
    "        des = 'Empoasca decipiens'\n",
    "    elif '3' in name:\n",
    "        des = 'Arboridia Creta'\n",
    "    elif '4' in name:\n",
    "        des = 'Arboridia Dalmazia'\n",
    "    return SeqRecord(Seq(s),id=name,description=des)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "9c64dd8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1A_H C 13 1A_L T 62 b\n",
      "1A_H A 11 1A_L G 62 b\n",
      "1D_H G 10 1D_L A 62 b\n",
      "4A_H A 54 4A_L C 62 b\n",
      "4C_H T 13 4C_L C 54 b\n",
      "4C_H A 13 4C_L G 62 b\n",
      "4D_H A 62 4D_L C 46 a\n",
      "4E_H A 62 4E_L C 59 a\n"
     ]
    }
   ],
   "source": [
    "# alignment of the two sequences (forward and reverse) using blast\n",
    "fs = [f for f in os.listdir(seq_folder) if f.endswith('.fa')] # collecting file names with sequencing information\n",
    "F = [x for x in fs if 'H' in x] # list of forward sequence files\n",
    "R = [x for x in fs if 'L' in x] # list of reverse sequence files\n",
    "l = list(zip(F,R)) # list of tuples (forward,reverse) of each sequence\n",
    "for tup in l:\n",
    "    s1,s2 = tup\n",
    "    output = NcbiblastnCommandline(query=seq_folder+s1, subject=seq_folder+s2, outfmt=5)()[0]\n",
    "    blast_result_record = NCBIXML.read(StringIO(output))\n",
    "    q_start = 9999\n",
    "    q_end = 0\n",
    "    for alignment in blast_result_record.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "            query = hsp.query\n",
    "            sbjct = hsp.sbjct\n",
    "            q_start = min(hsp.query_start,q_start)\n",
    "            q_end = max(hsp.query_end,q_end)\n",
    "            if '-' in hsp.query:\n",
    "                print(s1)\n",
    "    seq1 = get_sequence(seq_folder+s1,q_start-1,q_end)\n",
    "    with open(edit_folder+s1,'w') as handle:\n",
    "        SeqIO.write(seq1,handle,'fasta')\n",
    "    s_start = 9999\n",
    "    s_end = 0\n",
    "    for alignment in blast_result_record.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "            s_start = min(hsp.sbjct_end,s_start)\n",
    "            s_end = max(hsp.sbjct_start,s_end)\n",
    "        if '-' in hsp.sbjct:\n",
    "                print(s2)\n",
    "    seq2 = get_sequence(seq_folder+s2,s_start-1,s_end)\n",
    "    with open(edit_folder+s2,'w') as handle:\n",
    "        SeqIO.write(seq2,handle,'fasta')\n",
    "    cons = get_consensus(query, sbjct, q_start-1, s_start-1, seq1.id, seq2.id, qualdic, 35)\n",
    "    with open(output_folder+cons.id+'.fa','w') as handle:\n",
    "        SeqIO.write(cons,handle,'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "id": "507a7a0a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATAGCACCGGCCAAAACTGGCAAAGATAATAAAAGTAAAATTGCTGTAATTACTACTGACCATACAAATAATGGAATTCGATCTAAACTCAAACTAATTGAACGCATGTTGATAACTGTAGTAATAAAATTTACTGCTCCCAGGATAGATGAAATCCCAGCTAAATGTAATGAAAAGATTGCTAAATCTACTCTTGCCCCAGAATGGGCAATATTAGATGAAAGAGGGGGGTAAACTGTTCATCCGGTTCCAGCCCCTGTTTCAATCATAGATCTCGTTAGTAATAAAGTTAATGAAGGAGGTAAAAGTCAAAATCTTATATTATTTATTCGGGGAAAAGCTATATCAGGGGCCCCAATCATTAAAGGTAAAAGTCAATTACCAAAACCCCCGATCATAATCGGCATTACTATAAAGAAAATTATAATAAATGCATGAGATGTTACAATAACATTATAAATTTGATCATTGTTTAAAAAGACC\n",
      "||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||| |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||\n",
      "ATAGCACCGGCCAAAACTGGCAAAGATAATAAAAGTAAAATTGCTGTAATTACTACTGACCATACAAATAATGGAATTCGATCTAAACTCAAACTAATTGAACGCATGTTGATAACTGTAGTAATAAAATTTACTGCTCCCAGGATAGATGAAATCCCAGCTAAATGTAATGAAAAGATTGCTAAATCTACTCTTGCTCCAGAATGGGCAATATTAGATGAAAGAGGGGGGTAAACTGTTCATCCGGTTCCAGCCCCTGTTTCAATCATAGATCTCGTTAGTAATAAAGTTAATGAAGGAGGTAAAAGTCAAAATCTTATATTATTTATTCGGGGGAAAGCTATATCAGGGGCCCCAATCATTAAAGGTAAAAGTCAATTACCAAAACCCCCGATCATAATCGGCATTACTATAAAGAAAATTATAATAAATGCATGAGATGTTACAATAACATTATAAATTTGATCATTGTTTAAAAAGACC\n",
      "198 13 62\n",
      "336 11 62\n"
     ]
    }
   ],
   "source": [
    "output = NcbiblastnCommandline(query='D:/PhD/PhD.AES.Dsuz.Riccardo.Piccinno/Cicaline/my_code/sequences/1A_H.fa', \n",
    "                               subject='D:/PhD/PhD.AES.Dsuz.Riccardo.Piccinno/Cicaline/my_code/sequences/1A_L.fa', \n",
    "                               outfmt=5)()[0]\n",
    "blast_result_record = NCBIXML.read(StringIO(output))\n",
    "q_start = 9999\n",
    "q_end = 0\n",
    "for alignment in blast_result_record.alignments:\n",
    "    for hsp in alignment.hsps:\n",
    "        print(hsp.query)\n",
    "        print(hsp.match)\n",
    "        print(hsp.sbjct)\n",
    "        seq = hsp.query\n",
    "        ids = [i for i in range(len(hsp.match)) if hsp.match[i] != '|']\n",
    "        for el in ids:\n",
    "            a = hsp.query[el]\n",
    "            b = hsp.sbjct[el]\n",
    "            if a == 'A':\n",
    "                if b == 'G':\n",
    "                    seq = seq[:el] + 'R' + seq[el+1:]\n",
    "                elif b == 'T':\n",
    "                    seq = seq[:el] + 'W' + seq[el+1:]\n",
    "                elif b == 'C':\n",
    "                    seq = seq[:el] + 'M' + seq[el+1:]\n",
    "                else:\n",
    "                    seq = seq[:el] + 'N' + seq[el+1:]\n",
    "            if a == 'G':\n",
    "                if b == 'A':\n",
    "                    seq = seq[:el] + 'R' + seq[el+1:]\n",
    "                elif b == 'T':\n",
    "                    seq = seq[:el] + 'K' + seq[el+1:]\n",
    "                elif b == 'C':\n",
    "                    seq = seq[:el] + 'S' + seq[el+1:]\n",
    "                else:\n",
    "                    seq = seq[:el] + 'N' + seq[el+1:]\n",
    "            if a == 'C':\n",
    "                if b == 'G':\n",
    "                    seq = seq[:el] + 'S' + seq[el+1:]\n",
    "                elif b == 'T':\n",
    "                    seq = seq[:el] + 'Y' + seq[el+1:]\n",
    "                elif b == 'A':\n",
    "                    seq = seq[:el] + 'M' + seq[el+1:]\n",
    "                else:\n",
    "                    seq = seq[:el] + 'N' + seq[el+1:]\n",
    "            if a == 'T':\n",
    "                if b == 'G':\n",
    "                    seq = seq[:el] + 'K' + seq[el+1:]\n",
    "                elif b == 'A':\n",
    "                    seq = seq[:el] + 'W' + seq[el+1:]\n",
    "                elif b == 'C':\n",
    "                    seq = seq[:el] + 'Y' + seq[el+1:]\n",
    "                else:\n",
    "                    seq = seq[:el] + 'N' + seq[el+1:]\n",
    "            print(hsp.query_start + el, qualdic['1A_H'][hsp.query_start + el-1], qualdic['1A_L'][-(hsp.query_start +el)])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
